doc/sphinx/proofs/writing-proofs/inductive.rst

```diff
@@ -494,22 +495,22 @@ and an explanation of the underlying technique.
 
 .. seealso:: :tacn:`functional induction`
 
-.. tacn:: discriminate @term
-   :name: discriminate
+.. tacn:: discriminate {? @destruction_arg }
 
    This tactic proves any goal from an assumption stating that two
    structurally different :n:`@term`\s of an inductive set are equal. For
    example, from :g:`(S (S O))=(S O)` we can derive by absurdity any
    proposition.
```

Comment:
The behavior is quirky.  I expected `0 = 1` or `1 <> 1` would derive a contradiction whether in a hypothesis or the goal.  But I get this:

```
Goal 1 = 2 -> True.
discriminate.   (* No more goals *)

Goal True -> 1 = 2.
discriminate.   (* No primitive equality found.  I expected to see 1 = 2 replaced with False *)

Goal 1 <> 1 -> True.
discriminate.   (* No primitive equality found. *)

Goal True -> 1 <> 2.
discriminate.   (* No more goals -- but 1 <> 2 is a true statement, didn't expect it to work on this one *)
```

Seems to me this a usability issue, perhaps worth fixing.  It may break existing proofs, though.

Reply:
Maybe what you're missing is that `x <> y` is the same as `x = y -> False`.  The behavior is exactly what the doc states: if, after introducing all hypotheses, there is a contradictory equality in the context, then `discriminate` will prove any goal from it.  The tactic will only make progress if it solves the goal, so it's not going to replace any of your goals with `False` (this is not really useful behavior anyway).  Additionally, `discriminate` is about discriminating different things, not about instantiating hypotheses which posit that two equal things are not equal; it's not about instantiating hypotheses with reflexive proofs.

Perhaps you're thinking of `congruence`, which is a more general equality reasoning tactic?

Reply:
Yeah, I forgot about what's behind some of the basic Prop notations.  There should be a short section somewhere in the doc that explains those details (e.g. True, False, ->, <>, the things in Logic.v) because they seem important for users to know about.  Any idea where to put something like that?

No, not thinking about `congruence`.

Reply:
There's https://coq.inria.fr/refman/language/coq-library.html